drivers: spi: fix infinite loop with async spi and CONFIG_SPI_STM32_INTERRUPT on STM32

Yiffyi opened the pull request:
Fix #94282.

With `CONFIG_SPI_STM32_INTERRUPT` enabled, `LL_SPI_HALF_DUPLEX_*` transfer direction switch is already done in `spi_stm32_isr`. There's no need to perform completion check in a while loop, which breaks asynchronous transfer with infinite loop, due to the fact that `spi_context_wait_for_completion` always returns 0 during asynchronous transfer.

Yiffyi committed:
drivers: spi: stm32: fix infinite loop in interrupt async mode

Fix a infinite loop in STM32 spi driver when performing asynchronous
SPI transfer with CONFIG_SPI_STM32_INTERRUPT=y. A single call to
spi_context_wait_for_completion() is enough to get the transfer status,
and transfer direction switch for half duplex is already done
in spi_stm32_isr().

Signed-off-by: Yiffyi Jia <git@yif.fyi>

etienne-lms reviewed (commented):
@Yiffyi, could you add a brief description of the change made (why, how) in the commit message body? The commit header line is also a bit too long, could be shortened, e.g.:
```
drivers: spi: stm32: fix infinite loop in interrupt async mode

Fix the infinite loop in STM32 SPI driver when CONFIG_SPI_ASYNC=y and
CONFIG_SPI_STM32_INTERRUPT=y. When interrupt mode is not used,
a single call to spi_context_wait_for_completion() is enough to get the
transfer status.

Signed-off-by: ...
```
Fell free to rephrase that, I just tried to describe the change you currently make.

Yiffyi commented:
> @Yiffyi, could you add a brief description of the change made (why, how) in the commit message body? The commit header line is also a bit too long, could be shortened, e.g.:
> 
> ```
> drivers: spi: stm32: fix infinite loop in interrupt async mode
> 
> Fix the infinite loop in STM32 SPI driver when CONFIG_SPI_ASYNC=y and
> CONFIG_SPI_STM32_INTERRUPT=y. When interrupt mode is not used,
> a single call to spi_context_wait_for_completion() is enough to get the
> transfer status.
> 
> Signed-off-by: ...
> ```
> 
> Fell free to rephrase that, I just tried to describe the change you currently make.

OK; I've updated the commit message. BTW, I fixed the compile warning for unused variable in the latest commit.

etienne-lms reviewed (commented):
@Yiffyi, I don't really see where does `spi_stm32_isr()` handle the direction switch as mentioned in the commit message.
